contracts/infrastructure/WalletFactory.sol

```diff
@@ -222,4 +223,60 @@ contract WalletFactory is Owned, Managed {
         require(_guardian != (address(0)), "WF: guardian cannot be null");
         require(_version > 0, "WF: invalid _version");
     }
+
+    /**
+     * @notice Refunds the creation of the wallet when provided with a vild signature from the wallet owner.
+     * @param _wallet The wallet created
+     * @param _owner The owner address
+     * @param _refundAmount The amount to refund
+     * @param _refundToken The token to use for the refund
+     * @param _ownerSignature A signature from the wallet owner approving the refund amount and token. 
+     */
+    function validateAndRefund(
+        address _wallet,
+        address _owner,
+        uint256 _refundAmount,
+        address _refundToken,
+        bytes memory _ownerSignature
+    )
+        internal
+    {
+        bytes32 signedHash = keccak256(abi.encodePacked(
+                "\x19Ethereum Signed Message:\n32",
+                keccak256(abi.encodePacked(_refundAmount, _refundToken))
+            ));
+        address signer = Utils.recoverSigner(signedHash, _ownerSignature, 0);
+        if (signer == _owner) {
+            if (_refundToken == ETH_TOKEN) {
+				invokeWallet(_wallet, refundAddress, _refundAmount, "");
+            } else {
+                bytes memory methodData = abi.encodeWithSignature("transfer(address,uint256)", refundAddress, _refundAmount);
+                invokeWallet(_wallet, _refundToken, 0, methodData);
+            }
+        }
+    }
+
+    function invokeWallet(
+        address _wallet,
+        address _to,
+        uint256 _value,
+        bytes memory _data
+    )
+        internal
+        returns (bytes memory _res) 
+    {
+        bool success;
+        (success, _res) = _wallet.call(abi.encodeWithSignature("invoke(address,uint256,bytes)", _to, _value, _data));
+        if (success && _res.length > 0) { //_res is empty if _wallet is an "old" BaseWallet that can't return output values
```

Comment:
Given that (1) we know `_wallet` is not an old BaseWallet and (2) we don't care about the return value, this can be simplified to:
```
if(!success) {
            // solhint-disable-next-line no-inline-assembly
            assembly {
                returndatacopy(0, 0, returndatasize())
                revert(0, returndatasize())
            }
}
```

Reply:
Following @elenadimitrova comment we do care about the return value for ERC20 tokens who's `transfer` method returns a bool. But I've simplified the code to account for (1).

Reply:
Fine but note that the vast majority of tokens don't actually use the bool to indicate a failure but revert instead. That includes all the tokens you want to support for the refund.

Reply:
You're right. And even if that was not true, the only risk is to create a wallet when the refund fails because of a lack of funds... which should never happens if the backend monitors deposits correctly.

Reply:
I am checking some of the refund tokens we support and so far I see they all return `bool`, e.g. ZRX and BAT (which are defined as refund tokens in the gas estimation document). Can we be specific here as to what tokens we'll support? Despite that considering the minimal cost of this check and that it's part of ERC20 I would prefer to leave it.

Reply:
According to the [spec](https://www.notion.so/argenthq/Make-L1-wallets-paid-666c018f4413452d9503cf07c513022d) the supported refund tokens are ETH, USDC, DAI, USDT, which revert on failure. ZRX and BAT do indeed return `false` on failure.

Reply:
We are planning on removing ZRX and BAT from the transaction refund tokens, but in any-case for creation the refund tokens will be limited to ETH, DAI, USDC and USDT. These tokens either return `true` or revert. The added logic will literally never been used so I would prefer to not include it.

Reply:
And we can always revisit this choice in a coupe on month when do the next wallet upgrade.

Reply:
I don't understand why is it a problem to properly support ERC20 instead of relying on configuration of what refund tokens we support. Alternatively we should remove it from the Relayer logic as well although I would prefer to keep it as a safeguard.

Reply:
I don't feel strongly either way. I've re-added the returned bytes to be consistent with the Relayer logic.